[BUG] Bulk modify_tags M2M write rolls back silently under ASGI/psycopg3 — tags do not persist

### Description

## Environment
- **Paperless-ngx:** 2.20.15
- **Install:** Docker on Debian LXC (Proxmox)
- **Database:** PostgreSQL 18 (psycopg3)
- **Server:** Granian / ASGI (default)

## Summary
Selecting one or more documents in the document list and applying a tag via the bulk-edit toolbar silently fails to persist the tag. UI sometimes shows a success toast, but on page refresh the tag is not present.

The same operation **works correctly** when:
- Editing tags from the document detail view (`PATCH /api/documents/{id}/`)
- Setting correspondent or document type via bulk edit

Only **bulk tag assignment** is broken.

## Postgres-side evidence

With `log_statement = 'all'` and `log_line_prefix = '%t [%p] %a %v '`, a single bulk `modify_tags` request on document 1787 (adding tag 539) produces:

```
[186] vxid 40/15  BEGIN
[186] vxid 40/15  SELECT ... documents_document_tags WHERE document_id IN (1787) AND tag_id IN (539)
[186] vxid 40/15  INSERT INTO documents_document_tags VALUES (1787, 539) ON CONFLICT DO NOTHING
[187] vxid 44/9   SELECT tag_id FROM documents_document_tags WHERE document_id = 1787
[187] vxid 44/11  INSERT INTO auditlog_logentry ... changes='{"tags": [171, [171]]}'
```

Key observations:
1. The M2M INSERT and the auditlog INSERT for the **same** request execute on **two different backend processes / two different virtual transactions** (vxid 40/15 vs 44/11).
2. **No COMMIT is ever emitted for backend 186's transaction.**
3. Backend 187's audit log records `"tags": [171, [171]]` — i.e. "no change" — because the new M2M row isn't visible from its connection.
4. Direct DB query after the operation confirms the tag is not present:
```sql
   SELECT tag_id FROM documents_document_tags WHERE document_id = 1787;
   -- returns only [171] (Inbox)
```


Adding a relevant data point: checking my install's django_migrations history, the most recent migrations applied to my database were:
```
 app       |                    name                     |            applied            
----------------+---------------------------------------------+-------------------------------
 documents      | 0020_drop_celery_results                    | 2026-04-21 04:01:44.285741+00
 documents      | 0019_task_system_redesign                   | 2026-04-21 04:01:44.276537+00
 paperless      | 0009_alter_applicationconfiguration_options | 2026-04-09 04:02:02.34632+00
 paperless      | 0008_replace_skip_archive_file              | 2026-04-07 04:03:19.377844+00
 documents      | 0018_saved_view_simple_search_rules         | 2026-04-04 04:03:34.071139+00
```
The 0019_task_system_redesign + 0020_drop_celery_results pair landed on my install on April 21. This is the same area as the failing before_task_publish_handler / documents_paperlesstask write. Whichever PR introduced these migrations is a strong candidate for the regression — likely a good place for triagers to start.


## Concurrent symptom (likely related root cause)

Every bulk-tag request also produces:
```
[ERROR] [paperless.handlers] Creating PaperlessTask failed
...
psycopg.OperationalError: the connection is closed
at src/documents/signals/handlers.py:1133 in before_task_publish_handler
```

The `bulk_update_documents` Celery task itself dispatches and completes successfully. Only the `PaperlessTask` tracking row insert fails.

I tested patching `before_task_publish_handler` to write through an **isolated** `psycopg.connect()` — that eliminates the connection error but **does not** fix the tag-rollback issue. Confirms the two are separate manifestations of broken connection/transaction handling in the bulk_edit code path, not cause-and-effect.

## Hypothesis

The bulk_edit view's `modify_tags` path crosses an async/sync boundary mid-request under ASGI, splitting the M2M write and the post-write hooks run in different sync_to_async thread pools (likely due to thread_sensitive=False somewhere in the call chain), so each gets a separate thread-local connection.. Each thread gets its own DB connection from Django's thread-local connection registry. The first thread's transaction containing the M2M INSERT never commits.


Suspected fix area: wrap the entire `modify_tags` operation in `bulk_edit.py` (or its calling view in `views.py`) inside a single `transaction.atomic()` block, executed via one `sync_to_async(..., thread_sensitive=True)` call so all writes share one connection.



### Steps to reproduce

1. Fresh Paperless 2.20.15 + PostgreSQL.
2. Create a tag (e.g. id 539).
3. Select any document in the list view.
4. Use bulk-edit toolbar → "Tags" → add the tag → confirm.
5. Refresh — tag is not present. Postgres logs show INSERT without COMMIT, audit log shows no diff.


### Webserver logs

```bash
Running migrations:
  No migrations to apply.
[init-index] Checking search index...
[init-superuser] Not creating superuser
[init-checks] Running Django checks
Search index is up to date.
System check identified no issues (0 silenced).
[custom-init] /custom-cont-init.d doesn't exist, nothing to do
[init-complete] paperless-ngx docker container init completed in 14 seconds
[init-complete] Starting services
[svc-flower] Checking if we should start flower...
[svc-flower] Not starting flower
[INFO] Starting granian (main PID: 241)
[INFO][0m Listening at: http://:::8000
[INFO] Spawning worker-1 with PID: 260
celery beat v5.6.2 (recovery) is starting.
[2026-05-03 00:01:16,348] [INFO] [paperless.asgi] [init] Paperless-ngx version: v2.20.15
[2026-05-03 00:01:16,583] [INFO] [paperless.management.consumer] Processing existing files in /usr/src/paperless/consume
[2026-05-03 00:01:16,584] [INFO] [paperless.management.consumer] Watching /usr/src/paperless/consume using native file system events
 
 -------------- celery@c34a267784cc v5.6.2 (recovery)
--- ***** ----- 
-- ******* ---- Linux-6.17.13-2-pve-x86_64-with-glibc2.41 2026-05-03 00:01:17
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         paperless:0x77408fd6e600
- ** ---------- .> transport:   redis://broker:6379//
- ** ---------- .> results:     disabled://
- *** --- * --- .> concurrency: 1 (prefork)
-- ******* ---- .> task events: ON
--- ***** ----- 
 -------------- [queues]
                .> celery           exchange=celery(direct) key=celery
                

[tasks]
  . documents.bulk_edit.delete
  . documents.bulk_edit.restore_archive_serial_numbers_task
  . documents.signals.handlers.process_cf_select_update
  . documents.tasks.build_share_link_bundle
  . documents.tasks.bulk_update_documents
  . documents.tasks.check_scheduled_workflows
  . documents.tasks.cleanup_expired_share_link_bundles
  . documents.tasks.consume_file
  . documents.tasks.empty_trash
  . documents.tasks.index_optimize
  . documents.tasks.llmindex_index
  . documents.tasks.remove_document_from_llm_index
  . documents.tasks.sanity_check
  . documents.tasks.train_classifier
  . documents.tasks.update_document_content_maybe_archive_file
  . documents.tasks.update_document_in_llm_index
  . documents.workflows.webhooks.send_webhook
  . paperless_mail.mail.apply_mail_action
  . paperless_mail.mail.error_callback
  . paperless_mail.tasks.process_mail_accounts

[2026-05-03 00:01:17,410] [INFO] [celery.worker.consumer.connection] Connected to redis://broker:6379//
[2026-05-03 00:01:17,425] [INFO] [celery.apps.worker] celery@c34a267784cc ready.
__    -    ... __   -        _
LocalTime -> 2026-05-03 00:01:17
Configuration ->
    . broker -> redis://broker:6379//
    . loader -> celery.loaders.app.AppLoader
    . scheduler -> celery.beat.PersistentScheduler
    . db -> /usr/src/paperless/data/celerybeat-schedule.db
    . logfile -> [stderr]@%INFO
    . maxinterval -> 5.00 minutes (300s)
[2026-05-03 00:01:17,557] [INFO] [celery.beat] beat: Starting...
/usr/local/lib/python3.12/site-packages/django/core/handlers/asgi.py[0m:336: Warning: StreamingHttpResponse must consume synchronous iterators in order to serve them asynchronously. Use an asynchronous iterator instead.
  async for part in content:
[2026-05-03 00:02:32,511] [ERROR] [paperless.handlers] Creating PaperlessTask failed
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/django/db/backends/base/base.py", line 298, in _cursor
    return self._prepare_cursor(self.create_cursor(name))
                                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/backends/postgresql/base.py", line 429, in create_cursor
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/psycopg/connection.py", line 239, in cursor
    self._check_connection_ok()
  File "/usr/local/lib/python3.12/site-packages/psycopg/_connection_base.py", line 532, in _check_connection_ok
    raise e.OperationalError("the connection is closed")
psycopg.OperationalError: the connection is closed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/paperless/src/documents/signals/handlers.py", line 1133, in before_task_publish_handler
    PaperlessTask.objects.create(
  File "/usr/local/lib/python3.12/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/query.py", line 665, in create
    obj.save(force_insert=True, using=self.db)
  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 902, in save
    self.save_base(
  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 1008, in save_base
    updated = self._save_table(
              ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 1169, in _save_table
    results = self._do_insert(
              ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/base.py", line 1210, in _do_insert
    return manager._insert(
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/query.py", line 1873, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 1880, in execute_sql
    with self.connection.cursor() as cursor:
         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/backends/base/base.py", line 320, in cursor
    return self._cursor()
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/backends/base/base.py", line 297, in _cursor
    with self.wrap_database_errors:
         ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.12/site-packages/django/db/backends/base/base.py", line 298, in _cursor
    return self._prepare_cursor(self.create_cursor(name))
                                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/django/db/backends/postgresql/base.py", line 429, in create_cursor
    cursor = self.connection.cursor()
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/psycopg/connection.py", line 239, in cursor
    self._check_connection_ok()
  File "/usr/local/lib/python3.12/site-packages/psycopg/_connection_base.py", line 532, in _check_connection_ok
    raise e.OperationalError("the connection is closed")
django.db.utils.OperationalError: the connection is closed
[2026-05-03 00:02:32,519] [INFO] [celery.worker.strategy] Task documents.tasks.bulk_update_documents[ec1af19e-6b00-4f66-8df4-a47f6230e55f] received
[2026-05-03 00:02:32,743] [INFO] [celery.app.trace] Task documents.tasks.bulk_update_documents[ec1af19e-6b00-4f66-8df4-a47f6230e55f] succeeded in 0.2229173870291561s: None
```

### Paperless-ngx version

2.20.15

### Host OS

Debian 6.17.13-2-pve

### Installation method

Docker - official image

### System status

```json
{
    "pngx_version": "2.20.15",
    "server_os": "Linux-6.17.13-2-pve-x86_64-with-glibc2.41",
    "install_type": "docker",
    "storage": {
        "total": 38257754112,
        "available": 9440231424
    },
    "database": {
        "type": "postgresql",
        "url": "paperless",
        "status": "OK",
        "error": null,
        "migration_status": {
            "latest_migration": "documents.0020_drop_celery_results",
            "unapplied_migrations": []
        }
    },
    "tasks": {
        "redis_url": "redis://broker:6379",
        "redis_status": "OK",
        "redis_error": null,
        "celery_status": "OK",
        "celery_url": "celery@c34a267784cc",
        "celery_error": null,
        "index_status": "OK",
        "index_last_modified": "2026-05-03T00:04:33.220891+03:00",
        "index_error": null,
        "classifier_status": "OK",
        "classifier_last_trained": "2026-05-02T21:05:04.721147Z",
        "classifier_error": null,
        "sanity_check_status": "OK",
        "sanity_check_last_run": "2026-04-25T21:30:35.792260Z",
        "sanity_check_error": null,
        "llmindex_status": "DISABLED",
        "llmindex_last_modified": null,
        "llmindex_error": null,
        "summary": {
            "days": 30,
            "total_count": 2209,
            "pending_count": 0,
            "success_count": 2170,
            "failure_count": 39
        }
    },
    "websocket_connected": "OK"
}
```

### Browser

Safari

### Configuration changes

docker compose postgres-tika


### Please confirm the following

- [x] I believe this issue is a bug that affects all users of Paperless-ngx, not something specific to my installation.
- [x] This issue is not about the OCR or archive creation of a specific file(s). Otherwise, please see above regarding OCR tools.
- [x] I have already searched for relevant existing issues and discussions before opening this report.
- [x] I have updated the title field above with a concise description.

--- a/src/documents/bulk_edit.py
+++ b/src/documents/bulk_edit.py
@@ -218,7 +218,7 @@ def modify_tags(
     doc_ids: list[int],
     add_tags: list[int],
     remove_tags: list[int],
-) -> Literal["OK"] | Literal["ERROR"]:
+) -> Literal["OK"]:
     qs = Document.objects.filter(id__in=doc_ids).only("pk")
     affected_docs = list(qs.values_list("pk", flat=True))
     DocumentTagRelationship = Document.tags.through
@@ -237,39 +237,34 @@ def modify_tags(
         expanded_remove_tags.add(int(t.id))
         expanded_remove_tags.update(int(pk) for pk in t.get_descendants_pks())
 
-    try:
-        with transaction.atomic():
-            if expanded_remove_tags:
+    with transaction.atomic():
+        if expanded_remove_tags:
+            DocumentTagRelationship.objects.filter(
+                document_id__in=affected_docs,
+                tag_id__in=expanded_remove_tags,
+            ).delete()
+
+        to_create = []
+        if expanded_add_tags:
+            existing_pairs = set(
                 DocumentTagRelationship.objects.filter(
                     document_id__in=affected_docs,
-                    tag_id__in=expanded_remove_tags,
-                ).delete()
+                    tag_id__in=expanded_add_tags,
+                ).values_list("document_id", "tag_id"),
+            )
 
-            to_create = []
-            if expanded_add_tags:
-                existing_pairs = set(
-                    DocumentTagRelationship.objects.filter(
-                        document_id__in=affected_docs,
-                        tag_id__in=expanded_add_tags,
-                    ).values_list("document_id", "tag_id"),
+            to_create = [
+                DocumentTagRelationship(document_id=doc, tag_id=tag)
+                for doc in affected_docs
+                for tag in expanded_add_tags
+                if (doc, tag) not in existing_pairs
+            ]
+
+            if to_create:
+                DocumentTagRelationship.objects.bulk_create(
+                    to_create,
+                    ignore_conflicts=True,
                 )
-
-                to_create = [
-                    DocumentTagRelationship(document_id=doc, tag_id=tag)
-                    for doc in affected_docs
-                    for tag in expanded_add_tags
-                    if (doc, tag) not in existing_pairs
-                ]
-
-                if to_create:
-                    DocumentTagRelationship.objects.bulk_create(
-                        to_create,
-                        ignore_conflicts=True,
-                    )
-
-    except Exception as e:
-        logger.error(f"Error modifying tags: {e}")
-        return "ERROR"
 
     if affected_docs:
         bulk_update_documents.apply_async(

--- a/src/documents/bulk_edit.py
+++ b/src/documents/bulk_edit.py
@@ -218,7 +218,7 @@ def modify_tags(
     doc_ids: list[int],
     add_tags: list[int],
     remove_tags: list[int],
-) -> Literal["OK"]:
+) -> Literal["OK"] | Literal["ERROR"]:
     qs = Document.objects.filter(id__in=doc_ids).only("pk")
     affected_docs = list(qs.values_list("pk", flat=True))
     DocumentTagRelationship = Document.tags.through
@@ -267,14 +267,15 @@ def modify_tags(
                         ignore_conflicts=True,
                     )
 
-            if affected_docs:
-                bulk_update_documents.apply_async(
-                    kwargs={"document_ids": affected_docs},
-                    headers={"trigger_source": PaperlessTask.TriggerSource.SYSTEM},
-                )
     except Exception as e:
         logger.error(f"Error modifying tags: {e}")
         return "ERROR"
+
+    if affected_docs:
+        bulk_update_documents.apply_async(
+            kwargs={"document_ids": affected_docs},
+            headers={"trigger_source": PaperlessTask.TriggerSource.SYSTEM},
+        )
 
     return "OK"
 

--- a/src/documents/signals/handlers.py
+++ b/src/documents/signals/handlers.py
@@ -1122,7 +1122,6 @@ def before_task_publish_handler(
         return
 
     try:
-        close_old_connections()
         _, task_kwargs, _ = body
         task_id = headers["id"]
 

